refactoring to typescript.

oubenruing opened the pull request:
As described in issue #72

bzuillsmith committed:
minor version bump for mtext support

bzuillsmith committed:
update the readme to reflect recent changes

Commit:
fix bounding box

bzuillsmith committed:
Merge pull request #71 from bercht-a/bug/BoundingBoxFix

fix bounding box calculation

bzuillsmith committed:
Patch release bump

oubenruing committed:
♻️ Refactoring to Typescript.

oubenruing committed:
🎨 enhanced visibility of point painting

oubenruing committed:
🐛 fix the bug of AABB calculation

oubenruing committed:
⬆️ update Dependencies

oubenruing committed:
🔥 delete js files

bzuillsmith commented:
Wow, good stuff. I haven't taken a deep look yet. Did anything major change with the API? If so, I might merge it into a v2 branch and try adding some automated tests and release as a 2.0.0 or something. But if nothing changed with how you use it, then I guess it could just be a minor version change. Either way, I've been thinking about how to run tests on it.

oubenruing commented:
There are some minor changes to the API,  I think it is better to create a new branch.
I modified the webpack build script so that after the build, the original sample still works.
``` bash
# first, compile three-dxf
> npm install
> npm run build

# then install the sample's dependencies
> cd sample
> npm install

# go back to the root and run http-server to run the sample
> cd ..
> npm install -g http-server@0.9.0
> http-server .
# use `http-server -c-1 .` to prevent caching

```
## Here are some details of the changes

#### I did the following to refactor the code：
- Import OrbitControls from threejs and remove OrbitControls.js from the src folder.
- Put the tool functions (e.g. bspline.js, round10.js and THREEx.Math.xxx in index.js) into the utils.ts file.
- Split index.js into index.ts and dxf-painter.ts. index.ts does the work related to the initialization of threejs. dxf-painter.ts does the drawing of dxf-parser results.

#### And I made the following changes：

- Improved the way the three.js render function is called.
``` ts
// origin
    this.render = function () { renderer.render(scene, camera) };
    controls.addEventListener('change', this.render);
    this.render();
    controls.update();

// new 
   const animate = () => {
      requestAnimationFrame(animate);
      this.render();
      this.controls.update();
    };
   animate();
```
- Added raycaster to detect mouse and 3d object interaction. （When the mouse collides with a drawn object, the object turns red）
```ts
  private render() {
    this.raycaster.setFromCamera(this.pointer, this.camera);
    // calculate objects intersecting the picking ray
    this.intersects = this.raycaster.intersectObjects(
      this.scene.children,
      true
    );

    if (this.intersects.length > 0) {
      this.parent.style.cursor = 'pointer';
      if (this.INTERSECTED != this.intersects[0].object) {
        if (this.INTERSECTED)
          this.INTERSECTED.material.color.set(this.INTERSECTED.currentHex);

        this.INTERSECTED = this.intersects[0].object;
        this.INTERSECTED.currentHex = this.INTERSECTED.material.color;
        this.INTERSECTED.material.color = new THREE.Color(0xff0000);
      }
    } else {
      if (this.INTERSECTED)
        this.INTERSECTED.material.color.set(this.INTERSECTED.currentHex);
      this.parent.style.cursor = 'auto';
      this.INTERSECTED = null;
    }
    this.renderer.render(this.scene, this.camera);
  }
```
- Comment out functions that are not applicable.（Some functions that are not called）
- Enhanced visibility of point painting.
```ts
// private drawPoint(entity: IPointEntity, data: IDxf) {
  //   const color = this.getColor(entity, data);
  //   const geometry = new THREE.BufferGeometry();

  //   geometry.setAttribute(
  //     'position',
  //     new THREE.Float32BufferAttribute(
  //       [entity.position.x, entity.position.y, entity.position.z ?? 1],
  //       3
  //     )
  //   );

  //   const material = new THREE.PointsMaterial({
  //     size: 0.1,
  //     color: new THREE.Color(color),
  //   });
  //   const point = new THREE.Points(geometry, material);

  //   return point;
  // }

  // Draw a point like a big circle
  private drawPoint(entity: IPointEntity, data: IDxf) {
    const color = this.getColor(entity, data);
    const geometry = new THREE.CircleGeometry(2, 12);
    const material = new THREE.MeshBasicMaterial({ color: color });
    const point = new THREE.Mesh(geometry, material);
    point.position.set(
      entity.position.x ?? 0,
      entity.position.y ?? 0,
      entity.position.z ?? 1
    );
    point.name = 'point';
    point.type = 'point';
    return point;
  }
```
- Fix the bug of AABB （Axis-aligned Bounding Box） calculation.
```ts
// origin
var i, entity, obj, min_x, min_y, min_z, max_x, max_y, max_z;
    var dims = {
        min: { x: 0, y: 0, z: 0 },
        max: { x: 0, y: 0, z: 0 }
    };
    for (i = 0; i < data.entities.length; i++) {
        entity = data.entities[i];
        obj = drawEntity(entity, data);

        if (obj) {
            var bbox = new THREE.Box3().setFromObject(obj);
            if (isFinite(bbox.min.x) && (dims.min.x > bbox.min.x)) dims.min.x = bbox.min.x;
            if (isFinite(bbox.min.y) && (dims.min.y > bbox.min.y)) dims.min.y = bbox.min.y;
            if (isFinite(bbox.min.z) && (dims.min.z > bbox.min.z)) dims.min.z = bbox.min.z;
            if (isFinite(bbox.max.x) && (dims.max.x < bbox.max.x)) dims.max.x = bbox.max.x;
            if (isFinite(bbox.max.y) && (dims.max.y < bbox.max.y)) dims.max.y = bbox.max.y;
            if (isFinite(bbox.max.z) && (dims.max.z < bbox.max.z)) dims.max.z = bbox.max.z;
            scene.add(obj);
        }
        obj = null;
    }

// new

//After drawing
const dims = new THREE.Box3().setFromObject(this.scene);
```
